### PR TITLE
initial squashfs support

### PR DESCRIPTION
**Pull request by Gottox**
- [x] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

This adds support for mounting squashfs archives into radare2. It uses libsqsh's low level APIs.

```
~/src/radare2 (master ✗) build/binr/radare2/radare2 /tmp/gentoo-current.xz.sqfs
[0x00000000]> m /mnt squashfs 0
INFO: Mounted squashfs on /mnt at 0x0
[0x00000000]> ls /mnt/xfce-extra/xfdashboard
r Manifest
r metadata.xml
r xfdashboard-1.0.0.ebuild
[0x00000000]> 
```

This implementation depends on unreleased features of [sqsh-tools](https://github.com/Gottox/sqsh-tools), so I added a wrapfile to pull the right version. Once libsqsh-1.4 is released, I'll finalized this PR.

### CODE REVIEW

**Comment by trufae** [comment]
Does it needs to be in subprojects/? maybe its a good time to move v35 disassemblers, sdb and capstone there and use `acr-wrap` to support meson's wrap files.

**Comment by trufae** [comment]
Check the nocs pipeline

**Comment by Gottox** [comment]
> libr/fs/p/fs_squashfs.c:137:		R_LOG_ERROR ("squashfs: Unknown file type. This is a bug. Please report it.");

What should the lint error tell me?

**Comment by Gottox** [comment]
Ah, it complains about the "." at the end of the string.

**Comment by Gottox** [comment]
I tested listing files. But apparently I'm unable to actually open and read them. Is there something I can do to test that too?

**Comment by trufae** [comment]
You can use the ms command to get into the fs shell and use ls, cat, get..

**Comment by Gottox** [comment]
```
[0x00000000]> m /mnt squashfs 0
INFO: Mounted squashfs on /mnt at 0x0
[0x00000000]> ms
[/]> cat /mnt/x11-themes/zukini/Manifest
DIST zukini-20120817.zip 195930 BLAKE2B 78ba7fa7fe5d21add63f1f5028d1616ea8a9cfbe885ea1b373a60ebc6c9acc1d0b7798a585c958ba092f52b6fc05bcbab8550d7524828983c665060cfada689b SHA512 0ea42bc729100018acc16405283d41a10381bbcc411ea1b473880b7fdfe6785d6b7ae409959bdfb4f86c6081ae5fd102701efff793ea852c9c9913a03805e5f0
EBUILD zukini-20120817.ebuild 795 BLAKE2B d1bb86f5edf20b2e4fe6dd7b76531bcaae2c083f746aebd4108824d000b9ea1c9845266b61edb18efc2b7f29f3f918f6251d7ae473976606af79cd3fdc6fdda5 SHA512 1ecfe10844466fb86d1e2a10f9eaf467fe62d428d44923472fd6e96d11ae6a479346a6d7eab7b078ba5ed5a5e74a2951940924f8fc67557e8ec68afe4cf7ddb4
MISC metadata.xml 249 BLAKE2B faf4934e0656d147934ba57e6e2ae85b3037faaba88046bc669de45d0174931f9e973f3abe0a8c8899f563fcef238aea226fe41bd85c4da1aabb85366466d9ff SHA512 93217a1bfd512baa8421bb1b23c4c2535dde9eb3795a25c08cba2e4f56e312980fd21823522e357a7a9ab288f3bb76362d9b7b8156913da96ee2b2b8b845e35a
[/]> 
```

nice.

**Comment by Gottox** [comment]
v1.4.0 is out! https://github.com/Gottox/sqsh-tools/releases/tag/v1.4.0

**Comment by Gottox** [comment]
v1.4.0 is out! https://github.com/Gottox/sqsh-tools/releases/tag/v1.4.0

Ready for review!

**Review by trufae — CHANGES_REQUESTED** [review]
Everything looks good to me, the only missing part is the build system thing:

* decide if we want to use subprojects instead of shlr/ from now on
* integrate meson's wrap with acr-wrap to make them available from Make

File: libr/fs/p/fs_squashfs.c [hunk 1]
```diff
@@ -0,0 +1,255 @@
+/* radare - BSD2 - Copyright 2024 - Enno Boland */
+
+#include <r_fs.h>
+#include <sqsh.h>
+
+static char sqsh_to_r_type (enum SqshFileType type) {
```
[the remark this answers is not in the document]
Comment (Gottox) [reply libr/fs/p/fs_squashfs.c:6]: As there's a `.clang-format`, I adhered to the style described there.

**Comment by Gottox** [comment]
>     * decide if we want to use subprojects instead of shlr/ from now on
>     * integrate meson's wrap with acr-wrap to make them available from Make

My suggestion is to  get rid of the subprojects in this PR. Decoupling this PR from the discussion.

File: libr/fs/p/fs_squashfs.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (trufae) [reply libr/fs/p/fs_squashfs.c:6]: I dont think clang format can handle this. If you know how to do it i would be happy to get an update here

**Comment by trufae** [comment]
About the subprojects thing im fine to merge it like this i can add the make support later but we cant add tests if the thing is not working on one of the build systems

**Comment by Gottox** [comment]
Also note the following:

* I only included fs_squashfs into meson. I didn't bother to understand the configure style one.

File: libr/fs/meson.build [hunk 2]
```diff
@@ -48,6 +48,16 @@ else
...
 endif
 
+if use_libsqsh
+  r_fs_sources += [
+    'p/fs_squashfs.c'
+  ]
+  r_fs_deps += [ libsqsh_dep ]
+  if get_option('blob')
+    r_fs_static_deps += [ libsqsh_static_dep ]
+  endif
```
Comment (Gottox) [inline libr/fs/meson.build:58]: I think this should be removed too, shouldn't it?

**Comment by trufae** [comment]
The linter again missing space before ( in free

**Comment by trufae** [comment]
after some thoughs i think the best we can do is to move capstone and v35 disassemblers which are pulled from the interwebs at build time into subprojects and make the preconfigure script pull them using acr-wrap. i'll work on that after merging this. and then enable squashfs for make. this will be also good for offline builds as it will simplify some of the work done in flatpak by @prodrigestivill

**Comment by Gottox** [comment]
At the moment, libsqsh does not support windows. I'll look into it during the next few days.

File: libr/fs/meson.build (hunk 2 above)
Comment (trufae) [reply libr/fs/meson.build:58]: Its fine i think

**Review by trufae — COMMENTED** [review]
Isnt meson building on windows too?

**Comment by trufae** [comment]
Ready to merge then?

**Comment by Gottox** [comment]
Go for it!

**Comment by trufae** [comment]
Squashed!

**Comment by trufae** [comment]
Beautiful contrib! Thank you!

i’ll do the make and subprojects parts

**Comment by trufae** [comment]
I’ll also add squashfs to the fuzzing box and report you any bug i found

**Pull request merged**